xmppserver/src/main/java/org/jivesoftware/openfire/SessionManager.java

```diff
@@ -1613,98 +1579,110 @@ public int getSessionDetachTime() {
-            // Upon leaving a cluster, the server uses its non-clustered/default ID again. Here, all clustered IDs are replaced with the new identity.
-            final NodeID defaultNodeID = server.getDefaultNodeID();
-            final NodeID nodeID = server.getNodeID();
-            if ( !defaultNodeID.equals( nodeID ) ) // In more recent versions of Openfire, the ID does not change.
-            {
-                CacheUtil.replaceValueInMultivaluedCache( componentSessionsCache, nodeID, defaultNodeID );
-                CacheUtil.replaceValueInCache( multiplexerSessionsCache, nodeID, defaultNodeID );
-                CacheUtil.replaceValueInCache( incomingServerSessionsCache, nodeID, defaultNodeID );
-            }
-
-            // The local cluster node left the cluster.
-            //
-            // Determine what entities were available only on other cluster nodes than the local one.
-            // These need to be cleaned up, as they're no longer available to the local cluster node.
-
-            // All entities that were connected to other cluster nodes are no longer available.
-            CacheUtil.retainValueInMultiValuedCache( componentSessionsCache, defaultNodeID );
-            CacheUtil.retainValueInCache( multiplexerSessionsCache, defaultNodeID );
-            final Set<StreamID> removedStreamIDs = CacheUtil.retainValueInCache( incomingServerSessionsCache, defaultNodeID );
-
-            // Update the collection of 'sockets/sessions coming from the same remote hostname' as well as the collection of 'validated domains' to reflect the loss of incoming server sessions.
-            if ( !removedStreamIDs.isEmpty() )
-            {
-                Log.debug( "The local cluster node left the cluster. The incoming server sessions with IDs '{}' were living on one (or more) other cluster nodes, and are no longer available.",
-                           String.join( ", ", removedStreamIDs.stream().map( StreamID::getID ).collect( Collectors.toSet() ) ) );
-                unregisterIncomingServerSession( removedStreamIDs );
-            }
-
-            // Note that the cleanup of client state in this class is triggered by a call from RoutingTableImpl#leftCluster() to SessionManager#removeRemoteClientSession().
+        // The local cluster node left the cluster.
+        if (XMPPServer.getInstance().isShuttingDown()) {
+            // Do not put effort in restoring the correct state if we're shutting down anyway.
+            return;
         }
+
+        // Upon leaving a cluster, clustered caches are reset to their local equivalent (by the swap from the clustered
+        // cache implementation to the default cache implementation that's done in the implementation of
+        // org.jivesoftware.util.cache.CacheFactory.leftCluster). This means that they now hold no data (as a new cache
+        // has been created). Data that's available on the local node needs to be added again.
+        restoreCacheContent();
```

Comment:
Is it necessary to tell any listeners that previously available 'remote' sessions are no longer available?

Reply:
Yes, good catch, I've missed this here.

What I (tried to) do is look at the 'regular flow' that adds or removes entries in the cache. As that flow includes invoking event listeners, so should the corresponding actions that are done when entries are added or removed as a result of joining or leaving a cluster.

Reply:
In hindsight, I don't think this is needed after all: There are six caches in play here, but only the content of one of them goes accompanied by firing off event listeners (sessionInfoCache). However, when already running in a clustered environment, those events are never broadcasted over the cluster, so there shouldn't be a need to do so for all sessions that were gained/lost when joining or leaving a cluster either.

Reply:
I've now added code comment to most implementations, to explicitly define what the thoughts are on the need to invoke event listeners.